File: chess-proj/chesskiller.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <array>
#include <map>
#include <filesystem>
#include <chrono>
#include <iomanip>

#include "chess.hpp"
#include "json.hpp" 

using json = nlohmann::json;
namespace fs = std::filesystem;

// ================= 配置区域 =================
const std::string PGN_FILE_PATH = "dataset/lichess_db_standard_rated_2025-11.pgn";
const std::string OUTCOME_DIR = "outcome/chess_killer";
const std::string CHECKPOINT_FILE = OUTCOME_DIR + "/checkpoint_cpp.json";
const std::string FINAL_RESULT_FILE = OUTCOME_DIR + "/final_matrix.json";

const int MAX_TURN_INDEX = 75; 
const int SAVE_INTERVAL = 10000;
// ===========================================

struct TrackedPiece {
    std::string id;
    bool promoted = false;
    bool active = false; // 相当于 Python 中的 None 检查
};

// 用于暂存一局游戏中的击杀事件 (事务机制)
struct PendingEvent {
    std::string attacker_id;
    std::string victim_id;
    int turn_idx;
};

// 全局统计数据
struct GlobalStats {
    // Kill Matrix: AttackerID -> VictimID -> TurnCounts
    std::map<std::string, std::map<std::string, std::vector<int>>> kill_matrix;
    // Death Timeline: VictimID -> TurnCounts
    std::map<std::string, std::vector<int>> death_timeline;

    void commit_event(const PendingEvent& e) {
        // 1. 更新 Kill Matrix
        auto& turns = kill_matrix[e.attacker_id][e.victim_id];
        if (turns.empty()) turns.resize(MAX_TURN_INDEX + 1, 0);
        turns[e.turn_idx]++;

        // 2. 更新 Death Timeline
        auto& d_turns = death_timeline[e.victim_id];
        if (d_turns.empty()) d_turns.resize(MAX_TURN_INDEX + 1, 0);
        d_turns[e.turn_idx]++;
    }
};

class AdvancedTracker {
public:
    std::array<TrackedPiece, 64> board_map;

    AdvancedTracker() { initialize_ids(); }

    void initialize_ids() {
        // 清空棋盘
        for(auto& p : board_map) p = { "", false, false };
        
        const std::vector<char> files = {'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'};
        auto set = [&](chess::Square sq, std::string id) {
            board_map[sq.index()] = {id, false, true};
        };

        // White Init
        set(chess::Square::SQ_A1, "White_Rook_a");
        set(chess::Square::SQ_B1, "White_Knight_b");
        set(chess::Square::SQ_C1, "White_Bishop_c");
        set(chess::Square::SQ_D1, "White_Queen");
        set(chess::Square::SQ_E1, "White_King");
        set(chess::Square::SQ_F1, "White_Bishop_f");
        set(chess::Square::SQ_G1, "White_Knight_g");
        set(chess::Square::SQ_H1, "White_Rook_h");
        for (int i = 0; i < 8; ++i) {
            set(chess::Square(chess::File(i), chess::Rank::RANK_2), "White_Pawn_" + std::string(1, files[i]));
        }

        // Black Init
        set(chess::Square::SQ_A8, "Black_Rook_a");
        set(chess::Square::SQ_B8, "Black_Knight_b");
        set(chess::Square::SQ_C8, "Black_Bishop_c");
        set(chess::Square::SQ_D8, "Black_Queen");
        set(chess::Square::SQ_E8, "Black_King");
        set(chess::Square::SQ_F8, "Black_Bishop_f");
        set(chess::Square::SQ_G8, "Black_Knight_g");
        set(chess::Square::SQ_H8, "Black_Rook_h");
        for (int i = 0; i < 8; ++i) {
            set(chess::Square(chess::File(i), chess::Rank::RANK_7), "Black_Pawn_" + std::string(1, files[i]));
        }
    }

    std::string get_full_id(const TrackedPiece& p) {
        if (!p.active) return "";
        return p.id + (p.promoted ? "_promoted" : "");
    }
};

class KillerVisitor : public chess::pgn::Visitor {
public:
    chess::Board board;
    AdvancedTracker tracker;
    GlobalStats& stats;
    
    std::vector<PendingEvent> current_game_events;
    bool current_game_invalid = false;

    int games_processed_count = 0;
    int target_skip_count = 0;
    std::chrono::steady_clock::time_point start_time;

    KillerVisitor(GlobalStats& s, int start_count) : stats(s), games_processed_count(start_count), target_skip_count(start_count) {
        start_time = std::chrono::steady_clock::now();
        current_game_events.reserve(100);
    }

    void startPgn() override {
        if (games_processed_count < target_skip_count) {
            skipPgn(true);
            return;
        }
        board = chess::Board(); 
        tracker.initialize_ids();
        current_game_events.clear();
        current_game_invalid = false;
    }

    void header(std::string_view key, std::string_view value) override {}
    void startMoves() override {}

    void move(std::string_view move_str, std::string_view comment) override {
        if (current_game_invalid) return;

        chess::Move move;
        try {
            move = chess::uci::parseSan(board, move_str); 
        } catch (...) {
            current_game_invalid = true;
            skipPgn(true);
            return; 
        }
        process_move_logic(move);
    }

    void endPgn() override {
        games_processed_count++;

        if (!current_game_invalid && games_processed_count > target_skip_count) {
            for (const auto& e : current_game_events) {
                stats.commit_event(e);
            }
        }

        if (games_processed_count > target_skip_count && games_processed_count % SAVE_INTERVAL == 0) {
            print_progress();
            save_checkpoint();
        }
    }

    void print_progress() {
        auto now = std::chrono::steady_clock::now();
        std::chrono::duration<double> elapsed = now - start_time;
        double speed = (games_processed_count - target_skip_count) / elapsed.count();
        std::cout << "Processed: " << games_processed_count 
                  << " | Time: " << std::fixed << std::setprecision(1) << elapsed.count() << "s"
                  << " | Speed: " << speed << " g/s" << std::endl;
    }

    void save_checkpoint() {
        json j;
        j["games_processed"] = games_processed_count;
        j["current_pgn_offset"] = 0;
        
        // Checkpoint 结构保持嵌套
        j["stats"]["kill_matrix"] = stats.kill_matrix;
        j["stats"]["death_timeline"] = stats.death_timeline;

        std::string temp_path = CHECKPOINT_FILE + ".tmp";
        std::ofstream o(temp_path);
        o << j;
        o.close();
        try { fs::rename(temp_path, CHECKPOINT_FILE); } catch (...) {}
    }

private:
    void process_move_logic(chess::Move move) {
        auto from_sq = move.from();
        auto to_sq = move.to();
        
        // 异常防御
        if (!tracker.board_map[from_sq.index()].active) {
            board.makeMove(move);
            return; 
        }

        // =========================================================
        // 关键修正：这里必须使用值拷贝 (TrackedPiece)，不能用 auto&
        // 否则后续清空 from_sq 时，attacker_data 也会变为空！
        // =========================================================
        TrackedPiece attacker_data = tracker.board_map[from_sq.index()];
        std::string attacker_id = tracker.get_full_id(attacker_data);
        
        int turn_idx = board.fullMoveNumber() - 1; 
        if (turn_idx < 0) turn_idx = 0;
        if (turn_idx > MAX_TURN_INDEX) turn_idx = MAX_TURN_INDEX;

        // 1. Capture Logic
        if (board.isCapture(move)) {
            chess::Square victim_sq = to_sq;
            if (move.typeOf() == chess::Move::ENPASSANT) {
                victim_sq = chess::Square(to_sq.file(), from_sq.rank());
            }

            // 获取受害者数据 (值拷贝)
            TrackedPiece victim_data = tracker.board_map[victim_sq.index()];
            
            if (victim_data.active) {
                std::string victim_id = tracker.get_full_id(victim_data);
                current_game_events.push_back({attacker_id, victim_id, turn_idx});
                
                // 移除受害者
                tracker.board_map[victim_sq.index()] = {"", false, false};
            }
        }

        // 2. Move Tracker Logic
        // 处理易位 (Move Rook)
        if (move.typeOf() == chess::Move::CASTLING) {
            bool kingSide = to_sq > from_sq;
            chess::Color c = board.sideToMove();
            chess::Rank rank = (c == chess::Color::WHITE) ? chess::Rank::RANK_1 : chess::Rank::RANK_8;
            
            chess::Square rook_from = kingSide ? chess::Square(chess::File::FILE_H, rank) : chess::Square(chess::File::FILE_A, rank);
            chess::Square rook_to = kingSide ? chess::Square(chess::File::FILE_F, rank) : chess::Square(chess::File::FILE_D, rank);
            
            // 移动车 (先拷贝，再清除，再赋值)
            TrackedPiece rook_data = tracker.board_map[rook_from.index()];
            tracker.board_map[rook_from.index()] = {"", false, false};
            tracker.board_map[rook_to.index()] = rook_data;
        }

        // 移动主棋子 (Attacker)
        // 1. 清空原位置
        tracker.board_map[from_sq.index()] = {"", false, false};
        
        // 2. 处理升变标记
        if (move.typeOf() == chess::Move::PROMOTION) {
            attacker_data.promoted = true;
        }
        
        // 3. 放入新位置
        tracker.board_map[to_sq.index()] = attacker_data;

        // 3. Update Engine Board
        board.makeMove(move);

        // 4. Checkmate Logic
        chess::Movelist moves;
        chess::movegen::legalmoves(moves, board);
        
        if (moves.empty() && board.inCheck()) {
            // 获取此时位于 to_sq 的攻击者 (也就是刚才移动的棋子)
            // 这里使用引用是可以的，因为移动已经完成，位置已经确定
            auto& final_attacker_data = tracker.board_map[to_sq.index()];
            std::string final_attacker_id = tracker.get_full_id(final_attacker_data);
            
            std::string loser_color = (board.sideToMove() == chess::Color::WHITE) ? "White" : "Black";
            std::string victim_id = loser_color + "_King";

            current_game_events.push_back({final_attacker_id, victim_id, turn_idx});
        }
    }
};

int main() {
    if (!fs::exists(OUTCOME_DIR)) fs::create_directories(OUTCOME_DIR);

    GlobalStats stats;
    int start_games = 0;

    if (fs::exists(CHECKPOINT_FILE)) {
        std::cout << "Loading checkpoint..." << std::endl;
        try {
            std::ifstream i(CHECKPOINT_FILE);
            json j;
            i >> j;
            start_games = j["games_processed"];
            if (j.contains("stats")) {
                if (j["stats"].contains("kill_matrix"))
                    stats.kill_matrix = j["stats"]["kill_matrix"].get<std::map<std::string, std::map<std::string, std::vector<int>>>>();
                if (j["stats"].contains("death_timeline"))
                    stats.death_timeline = j["stats"]["death_timeline"].get<std::map<std::string, std::vector<int>>>();
            }
            std::cout << "Resuming from Game " << start_games << std::endl;
        } catch (...) {
            start_games = 0;
        }
    }

    std::ifstream pgn_file(PGN_FILE_PATH, std::ios::binary);
    if (!pgn_file.is_open()) {
        std::cerr << "PGN file not found." << std::endl;
        return 1;
    }

    KillerVisitor visitor(stats, start_games);
    chess::pgn::StreamParser parser(pgn_file);
    parser.readGames(visitor);

    // --- 最终保存 ---
    // 确保这里是扁平结构，直接输出 kill_matrix 和 death_timeline
    json j_final;
    j_final["kill_matrix"] = stats.kill_matrix;
    j_final["death_timeline"] = stats.death_timeline;
    
    std::cout << "Saving final results..." << std::endl;
    std::ofstream o(FINAL_RESULT_FILE);
    o << j_final.dump(4);
    o.close();

    std::cout << "Done!" << std::endl;
    return 0;
}

File: chess-proj/debug/chesskiller.py
import chess
import chess.pgn
import json
import os
import time
from collections import defaultdict

# ================= 配置区域 =================
# PGN_FILE_PATH = "dataset/lichess_db_standard_rated_2025-11.pgn"
PGN_FILE_PATH = "../dataset/experiment.pgn"
OUTCOME_DIR = "../outcome/chess_killer"
CHECKPOINT_FILE = os.path.join(OUTCOME_DIR, "checkpoint_v3_test.json")
FINAL_RESULT_FILE = os.path.join(OUTCOME_DIR, "final_matrix_py.json")

# 统计上限回合数 (索引0-74对应1-75回合，索引75对应76+回合)
MAX_TURN_INDEX = 75 
SAVE_INTERVAL = 10000  # 每处理多少局保存一次
# ===========================================

class AdvancedTracker:
    def __init__(self):
        self.board_map = [None] * 64
        self.initialize_ids()

    def initialize_ids(self):
        files = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
        # 初始化逻辑与之前相同，确保每个棋子有唯一ID
        # White
        self.board_map[chess.A1] = {"id": "White_Rook_a", "promoted": False}
        self.board_map[chess.B1] = {"id": "White_Knight_b", "promoted": False}
        self.board_map[chess.C1] = {"id": "White_Bishop_c", "promoted": False}
        self.board_map[chess.D1] = {"id": "White_Queen", "promoted": False}
        self.board_map[chess.E1] = {"id": "White_King", "promoted": False}
        self.board_map[chess.F1] = {"id": "White_Bishop_f", "promoted": False}
        self.board_map[chess.G1] = {"id": "White_Knight_g", "promoted": False}
        self.board_map[chess.H1] = {"id": "White_Rook_h", "promoted": False}
        for i, f in enumerate(files):
            self.board_map[chess.A2 + i] = {"id": f"White_Pawn_{f}", "promoted": False}

        # Black
        self.board_map[chess.A8] = {"id": "Black_Rook_a", "promoted": False}
        self.board_map[chess.B8] = {"id": "Black_Knight_b", "promoted": False}
        self.board_map[chess.C8] = {"id": "Black_Bishop_c", "promoted": False}
        self.board_map[chess.D8] = {"id": "Black_Queen", "promoted": False}
        self.board_map[chess.E8] = {"id": "Black_King", "promoted": False}
        self.board_map[chess.F8] = {"id": "Black_Bishop_f", "promoted": False}
        self.board_map[chess.G8] = {"id": "Black_Knight_g", "promoted": False}
        self.board_map[chess.H8] = {"id": "Black_Rook_h", "promoted": False}
        for i, f in enumerate(files):
            self.board_map[chess.A7 + i] = {"id": f"Black_Pawn_{f}", "promoted": False}

    def get_full_id(self, piece_data):
        if piece_data is None: return None
        suffix = "_promoted" if piece_data['promoted'] else ""
        return f"{piece_data['id']}{suffix}"

def get_turn_index(board_fullmove):
    # 回合数从1开始。Index 0 对应 Turn 1.
    # 如果超过75回合，归入 index 75
    idx = board_fullmove - 1
    if idx < 0: idx = 0 # 防御性编程
    if idx > MAX_TURN_INDEX: idx = MAX_TURN_INDEX
    return idx

def update_stats(stats, attacker_id, victim_id, turn_idx):
    """
    更新全局统计数据 (3D Matrix)
    """
    # 1. 更新 Kill Matrix: attacker -> victim -> turn
    if attacker_id not in stats["kill_matrix"]:
        stats["kill_matrix"][attacker_id] = {}
    
    if victim_id not in stats["kill_matrix"][attacker_id]:
        # 初始化一个长度为 MAX_TURN_INDEX + 1 的数组
        stats["kill_matrix"][attacker_id][victim_id] = [0] * (MAX_TURN_INDEX + 1)
    
    stats["kill_matrix"][attacker_id][victim_id][turn_idx] += 1

    # 2. 更新 Death Timeline: victim -> turn
    if victim_id not in stats["death_timeline"]:
        stats["death_timeline"][victim_id] = [0] * (MAX_TURN_INDEX + 1)
    
    stats["death_timeline"][victim_id][turn_idx] += 1

def process_game_moves(game, stats):
    board = game.board()
    tracker = AdvancedTracker()

    for move in game.mainline_moves():
        from_sq = move.from_square
        to_sq = move.to_square
        
        attacker_data = tracker.board_map[from_sq]
        if attacker_data is None: 
            board.push(move)
            continue # 数据异常跳过

        attacker_id = tracker.get_full_id(attacker_data)
        turn_idx = get_turn_index(board.fullmove_number)

        # --- 1. 处理吃子 (Physical Capture) ---
        if board.is_capture(move):
            victim_data = None
            
            if board.is_en_passant(move):
                # 吃过路兵：受害者在旁边
                offset = -8 if board.turn == chess.WHITE else 8
                victim_sq = to_sq + offset
                victim_data = tracker.board_map[victim_sq]
                # 移除被吃掉的兵
                tracker.board_map[victim_sq] = None
            else:
                # 普通吃子
                victim_data = tracker.board_map[to_sq]

            if victim_data:
                victim_id = tracker.get_full_id(victim_data)
                update_stats(stats, attacker_id, victim_id, turn_idx)

        # --- 2. 移动 Tracker 逻辑 ---
        # 处理易位
        if board.is_castling(move):
            if to_sq == chess.G1:   # White K-side
                tracker.board_map[chess.F1] = tracker.board_map[chess.H1]
                tracker.board_map[chess.H1] = None
            elif to_sq == chess.C1: # White Q-side
                tracker.board_map[chess.D1] = tracker.board_map[chess.A1]
                tracker.board_map[chess.A1] = None
            elif to_sq == chess.G8: # Black K-side
                tracker.board_map[chess.F8] = tracker.board_map[chess.H8]
                tracker.board_map[chess.H8] = None
            elif to_sq == chess.C8: # Black Q-side
                tracker.board_map[chess.D8] = tracker.board_map[chess.A8]
                tracker.board_map[chess.A8] = None
        
        # 移动棋子
        tracker.board_map[to_sq] = attacker_data
        tracker.board_map[from_sq] = None

        # 处理升变
        if move.promotion:
            tracker.board_map[to_sq]['promoted'] = True

        # 在逻辑棋盘落子 (必须先 update tracker 再 push，否则判断过路兵位置会麻烦)
        board.push(move)

        # --- 3. 处理将杀 (Checkmate as a Kill) ---
        if board.is_checkmate():
            # 此时 attacker_data 已经在 to_sq (如果不是升变)
            # 注意：如果刚才这步是升变，Attacker ID 应该变成 promoted 吗？
            # 逻辑上：兵到底线变后后将杀，应该是 "White_Pawn_a_promoted" 杀了王。
            # 我们重新获取一次 id，确保状态最新
            final_attacker_data = tracker.board_map[to_sq]
            final_attacker_id = tracker.get_full_id(final_attacker_data)
            
            # 受害者是对方的王
            opponent_king_color = "Black" if board.turn == chess.WHITE else "White" 
            # board.turn 在 push 后已经变了，如果现在是 White 走，说明刚才 Black 杀的
            # 哎呀，board.push(move) 之后，轮次已经交给对方了。
            # 如果 board.is_checkmate() 为真，说明当前轮次方无法动弹，被上一手方杀死了。
            
            loser_color = "White" if board.turn == chess.WHITE else "Black"
            victim_id = f"{loser_color}_King" 
            
            # 记录虚空击杀
            update_stats(stats, final_attacker_id, victim_id, turn_idx)

def run_advanced_analysis():
    if not os.path.exists(OUTCOME_DIR):
        os.makedirs(OUTCOME_DIR)

    # 数据结构初始化
    stats = {
        "kill_matrix": {},    # 3D: Attacker -> Victim -> [Turn Counts]
        "death_timeline": {}  # 2D: Victim -> [Turn Counts]
    }
    games_processed = 0
    file_offset = 0

    # 断点续传
    if os.path.exists(CHECKPOINT_FILE):
        print(f"Loading checkpoint: {CHECKPOINT_FILE}")
        try:
            with open(CHECKPOINT_FILE, 'r') as f:
                data = json.load(f)
                stats = data.get("stats", stats)
                games_processed = data.get("games_processed", 0)
                file_offset = data.get("current_pgn_offset", 0)
            print(f"Resuming from Game {games_processed}, Offset {file_offset}")
        except Exception as e:
            print(f"Checkpoint load failed: {e}. Starting fresh.")

    if not os.path.exists(PGN_FILE_PATH):
        print("PGN file not found.")
        return

    print("Starting Advanced Analysis...")
    start_time = time.time()

    with open(PGN_FILE_PATH) as pgn_file:
        pgn_file.seek(file_offset)
        
        while True:
            try:
                # 记录这局开始前的偏移量，以便保存
                offset_before_game = pgn_file.tell()
                game = chess.pgn.read_game(pgn_file)
            except Exception:
                continue

            if game is None:
                break

            process_game_moves(game, stats)
            games_processed += 1

            if games_processed % SAVE_INTERVAL == 0:
                elapsed = time.time() - start_time
                print(f"Processed: {games_processed} | Time: {elapsed:.1f}s | Speed: {games_processed/elapsed:.1f} g/s")
                
                # 获取读完后的 offset
                current_offset = pgn_file.tell()
                
                checkpoint_data = {
                    "games_processed": games_processed,
                    "current_pgn_offset": current_offset,
                    "stats": stats
                }
                
                # 原子写入
                temp_path = CHECKPOINT_FILE + ".tmp"
                with open(temp_path, 'w') as f:
                    json.dump(checkpoint_data, f) # 这里不indent以节省空间和IO
                os.replace(temp_path, CHECKPOINT_FILE)

    # Final Save
    with open(FINAL_RESULT_FILE, 'w') as f:
        json.dump(stats, f, indent=4) # 最终结果格式化以便人类阅读
    print(f"Done! Saved to {FINAL_RESULT_FILE}")

if __name__ == "__main__":
    run_advanced_analysis()